### ISSUE TITLE
Pox error and packet loss when running openflow discovery and spanning tree on large fat tree topologies

### ISSUE THREAD

**Issue by gmoleski** [noxrepo/pox#139]
Hi whenever I am running pox with this command "pox.py forwarding.l2_multi openflow.discovery openflow.spanning_tree --no-flood --hold-down" and a fat tree topology with 50-80 hosts I can't make them communicate. Some can communicate but most not. I think this is mostly due to link timeout, however when I set a larger link timeout still the problem isn't entirely solved. Is there a way to get around this?

Some of the errors I also get are:
ERROR:openflow.of_01:[00-00-00-00-00-20 457] OpenFlow Error:
[00-00-00-00-00-20 457] Error: header:
[00-00-00-00-00-20 457] Error:   version: 1
[00-00-00-00-00-20 457] Error:   type:    1 (OFPT_ERROR)
[00-00-00-00-00-20 457] Error:   length:  36
[00-00-00-00-00-20 457] Error:   xid:     112392
[00-00-00-00-00-20 457] Error: type: OFPET_BAD_REQUEST (1)
[00-00-00-00-00-20 457] Error: code: OFPBRC_BUFFER_UNKNOWN (8)
[00-00-00-00-00-20 457] Error: datalen: 24
[00-00-00-00-00-20 457] Error: 0000: 01 0d 00 18 00 01 b7 08  00 00 03 4f 0000 08   |...........O....|
[00-00-00-00-00-20 457] Error: 0010: 00 00 00 08 ff fb 00 00                        |........        |

And also l2_multi throws the error "packet arrived without flow".

Thanks.

**Comment by MurphyMc** [comment]
What version of POX?

**Comment by gmoleski** [comment]
I believe I am using version 2.0 of branch carp

**Comment by MurphyMc** [comment]
A couple quick things to try:
1) Upgrade to POX dart (a more recent version is always the first thing I try)
2) Pass --eat-early-packets to discovery

**Comment by gmoleski** [comment]
Thanks for the information but after running it with the latest version of dart I noticed that it is worse than carp. In dart after some time (I imagine link timeout) the nodes fail completely and errors are continuously thrown in the pox cmd. In carp I get some packet losses at the start and then when I re-run my traffic generator packet losses are decreasing but that is if I get the correct link timeout parameter for a certain topology. I will try to play more with the parameters and let you know. Thanks.

**Comment by mvanotti** [comment]
Hi! I'm having this same problem! is there any way to fix it?

**Comment by MurphyMc** [comment]
Post the POX log and a description of what you're doing.

**Comment by mvanotti** [comment]
Hi!
I'm running mininet with a Fat Tree Topology (TreeTopo depth=3, fanout=4)
Once I start the experiment, I try to create a TCP connection between every pair of nodes and exchange one small message (< 100bytes) every 1 second, for thirty seconds).

The problem I am having is that some of the hosts can connect, and in the pox log, I'm getting:

```
ERROR:openflow.of_01:[00-00-00-00-00-01 6] OpenFlow Error:
[00-00-00-00-00-01 6] Error: header: 
[00-00-00-00-00-01 6] Error:   version: 1
[00-00-00-00-00-01 6] Error:   type:    1 (OFPT_ERROR)
[00-00-00-00-00-01 6] Error:   length:  36
[00-00-00-00-00-01 6] Error:   xid:     60475
[00-00-00-00-00-01 6] Error: type: OFPET_BAD_REQUEST (1)
[00-00-00-00-00-01 6] Error: code: OFPBRC_BUFFER_UNKNOWN (8)
[00-00-00-00-00-01 6] Error: datalen: 24
[00-00-00-00-00-01 6] Error: 0000: 01 0d 00 18 00 00 ec 3b  00 00 05 11 00 05 00 08   |.......;........|
[00-00-00-00-00-01 6] Error: 0010: 00 00 00 08 ff fb 00 00                            |........        |
```

and

```
INFO:packet:(ipv6) warning IP packet data incomplete (114 of 149)
INFO:packet:(dns) parsing questions: next_question: truncated
```

My end goal is to run an internet-like topology using mininet, with at least, 100 nodes.

**Comment by MurphyMc** [comment]
Post the whole log and the commandline you're using.

**Comment by mvanotti** [comment]
Sorry about that! I uploaded everything to a gist [here](https://gist.github.com/mvanotti/7a979fede8e8f62e21b064a6f76e23e6). In the gist you will find:
- The pox log.
- The mininet log.
- The mininet script that I'm running
- The source code of the networking program that I'm running.

I've run Pox from the `carp` branch with the following parameters:
`./pox.py forwarding.l2_learning`

Please let me know if there's anything else I can do.

**Comment by MurphyMc** [comment]
That commandline definitely won't work.  See, the following POX FAQ entry, for example: https://openflow.stanford.edu/display/ONL/POX+Wiki#POXWiki-DoesPOXsupporttopologieswithloops%3F

I'd suggest upgrading to POX eel, and then starting with:
pox.py forwarding.l2_learning openflow.discovery --eat-early-packets openflow.spanning_tree --no-flood --hold-down

or 

pox.py forwarding.l2_multi openflow.discovery --eat-early-packets openflow.spanning_tree --no-flood --hold-down

You may have some luck tweaking the discovery timeouts too.

**Comment by mvanotti** [comment]
Hi! Thanks for your answer. 

I've re-ran the test with eel code and the first command you send. It still doesn't work :(. [Here's](https://gist.github.com/mvanotti/55acb3ecc52786c8d75c5479f948b155) the full log.

When you mean the discovery timeouts, do you mean my own tcp connections timeouts or something related to pox?

Thanks!

**Comment by MurphyMc** [comment]
A first note is that I think your life will improve a bit if you disable IPv6.

You might have better luck with the second command.  l2_learning with a spanning tree is a pretty silly way to run a network with a whole lot of loops (like a fat tree) anyway.

I believe what's going on here is that packets are getting caught in a loop which is causing discovery to fail.  I don't immediately know why this is happening, quite possibly something to do with the event handler priorities... the --eat-early-packets option is meant to prevent this, but it apparently isn't.  Try adding something like the following to the very start of the PacketIn handler in the forwarding component you're using (e.g., l2_learning):
`if (time.time() - event.connection.connect_time) <= 1.5 * core.openflow_discovery.send_cycle_time: return
`

See if that helps.  The idea is to make sure it doesn't do any forwarding until discovery has had a chance to discover.  (Again, this is what --eat-early-packets is meant to make happen, but this is a more direct way of doing it.)

By discovery timeouts, I meant the --link-timeout parameter to openflow.discovery.  Increasing it may help things.  It's currently coupled to how fast discovery cycles (sends packets out all the ports), which isn't fundamental... you could make a somewhat more "forgiving" version by keeping the cycle time shorter but making the timeouts longer.

**Comment by mvanotti** [comment]
Hi! Thanks for your answer!

I've tried the other method and it fails too. [Here is the complete log](https://gist.github.com/mvanotti/7003810df6d6d1196e19f0f3f0ac1703). I ran this command:

`./pox.py forwarding.l2_multi openflow.discovery --eat-early-packets openflow.spanning_tree --no-flood --hold-down`

I'm not sure I follow. How does the TreeTopo contains any loops? It is just a tree, no loops. Why would it have any problem? (I opted for this topology because I wanted to discard other possible problems relating loops in the topology).

Maybe I'm doing something that is wrong. I want to simulate a big network (>100 nodes) using mininet and pox. Maybe running everything in a big tree topo is not the correct way.

**Comment by MurphyMc** [comment]
Ah.  You'd said above that it was a fat tree, which has lots of loops.  If it's just a plain tree, then that's not the problem.  Maybe the problem is just too much going on at once.

It looked like maybe you're running a ping test between everything before starting your test.  Is that right?  What are the results from that?  Are the pings working?

Since there are no loops, there's no need for discovery/spanning_tree.  Try:

./pox.py forwarding.l2_pairs

**Comment by mvanotti** [comment]
Hi!

I too think the problem is "too much going on at once". But I don't know if that's solvable.
I always create 64 nodes, but I choose randomly 20, 32, and 64 of them.
For 20, if I run a ping with the 20 nodes, I don't usually get errors.
For 32, if I run a ping between the 32 nodes, I sometimes get errors.
For 64, running a ping with the 64 nodes, takes ~40 minutes and the few times I tried it, I got errors.

In all the experiments I sleep for 10 seconds before starting the ping, and then for 10 seconds again after the ping. 

Running `./pox.py forwarding.l2_pairs` also got me errors.

PS: I was under the impression that Fat Tree didn't have loops either, at least not the Fat Tree used in Maxinet. But the end goal is to test it with a topology with lots of loops, so..

**Comment by mvanotti** [comment]
I've just tried to run [this example from mininet](https://github.com/mininet/mininet/blob/master/examples/treeping64.py) specifying the POX controller, and I got packet loss during the pings running with `./pox.py forwarding.l2_pairs` and also with `./pox.py forwarding.l2_multi openflow.discovery --eat-early-packets openflow.spanning_tree --no-flood --hold-down`

**Comment by MurphyMc** [comment]
When people talk about fat trees (at least in my experience in networking), they're almost always referring to the topology discussed by Al-Fares et al., which are also (probably more properly) called "folded Clos networks".  Lots of loops.

The fact that you eventually want to run on a topology with lots of loops will hopefully be fine considering the problem you're seeing apparently doesn't have to do with loops.  It's definitely easier to use a no-loop network to start off.

Try modifying the example to run the ping tests more than once with a few seconds between.  Does it eventually get to 100%?

**Comment by mvanotti** [comment]
Sorry if I was not clear enough. When I ran the test with 64 nodes, none of the packets got lost, but I still got connection errors on my tests.

In the example above, I just got some errors the first time I ran it. Each full ping test takes like ~40 minutes with my computer using 100% cpu

**Comment by MurphyMc** [comment]
Okay, let's get on the same page.

First, pull _the very latest version of POX eel_.  I just pushed a fix to a dumb bug.

Modify the mininet treeping64 example.  You may have already done some of this.
1. Comment out the UserSwitch line (so that it only runs using the OVS kernel switch).
2. Adjust it so that it runs a POX controller (probably just using RemoteController).
3. Replace the `result = network.run( network.pingAll )` so that we can insert a pause.  Should be something like...

```
network.start()
time.sleep(20) # Give discovery time to cycle
result = network.pingAll()
network.stop()
```

Then try the modified example with "forwarding.l2_pairs" and "openflow.discovery forwarding.l2_multi".  Both of these work fine here.  Takes a couple minutes to complete in either case.

**Comment by mvanotti** [comment]
Hi!  Thank you very much for taking the time to test this out.

I can't see your push on github.

Ugh, I can't reproduce it now. But I **swear** that it wasn't working yesterday.

The other test I have been doing still fails however :(. (It's more complex than a simple ping, it requires establishing TCP Connections between nodes).

**Comment by mvanotti** [comment]
I was able to run 64 nodes in my computer! For this test I did a full ping between all pair of nodes and then waited 20 seconds before and after the ping.

It seems that the problem is then related to timing. Do you know what can be causing this? Is it possible to be sure when can I start sending packets ?

**Comment by MurphyMc** [comment]
Sorry, I pushed it to my fork, not upstream.  I've pushed it upstream now too.  I think you should probably upgrade to it.  The issue fixed here could definitely mess up your experiments in a somewhat-hard-to-reproduce-reliably way (e.g., depending on when Mininet and POX start up relative to each other).

And right; I understood that your real test was more complex than a simple ping.  I wanted to start with a simpler test and work our way up.

Yes, I also believe it has to do with timing, but I don't know exactly what aspect yet.  There are at least two possibilities.
1. If using discovery/spanning_tree, some of the options disable flooding and attempt to "eat" (discard) packets before discovery has had a chance to discover all links.  These approaches are for safety -- they drop packets rather than flood them before the topology is known (or before the spanning tree is set up), since before that... flooding them may result in packets caught in loops.  So your experiment will fail until discovery's "cycle time" has elapsed.  As mentioned above, this is coupled to discovery's link-timeout parameter (it needn't be so strongly coupled as it is, though). IIRC, the cycle time is about half of the timeout time, so if the timeout time is 10 seconds (the default, I think), then discovery won't have finished until some time after 5 seconds... before that, packet forwarding will likely not work.
2. Reactive control bottleneck.  If you are using an L2 "learning switch" type controller like l2_pairs, say, 1000 nodes in a tree, then the first time a host tries to open a TCP connection to another host X hops away, it will cause 1000 packet-in messages (for the ARP), and then X more packet-ins for the ARP reply.  If many hosts are trying to start communicating at the exact same time, it may be considerably more than this.  And each of these needs to get sent to the controller, the controller needs to process it, and the controller will probably send back a flow-mod, which then needs to be sent to and processed by the switch.  This may bottleneck at several places.  The controller's CPU performance may be one such place.  There are things one can do to improve the situation in POX (e.g., use the --unthreaded-sh option, and use PyPy instead of stock Python), but that just changes where things break, not that they break.  Exact-match controllers like l2_learning will do even worse because the installed table entries won't be reused across connections.  There are a number of things one can do to help resolve this, which generally involve not doing things reactively or at least not doing them _as_ reactively.  Or at least not doing them reactively _at the controller_.  POX has a few examples which change things up in this regard (though the assumption is generally that people will write controllers using POX to fit their exact needs, not that POX ships with best-for-everybody turnkey solutions).  l2_nx_self_learning, for example, does learning within the switch using Nicira/OVS extensions.  topo_proactive uses topologically meaningful addresses to avoid reactive learning altogether (e.g., much more like traditional routing).

If the problem is (1), it's just a matter of waiting a bit to start your test.

If the problem is (2), you could do a number of things.  For example, try to improve the performance (e.g., using --unthreaded-sh), or use a different forwarding/routing approach (several possibilities included with POX, or write your own... there are many "better" ways to do it than POX includes, particularly if you're willing to use stuff beyond plain OpenFlow 1.0), or you could "cheat" a bit.  This last one can be pretty effective.  The idea is to use learning, but to "prime" learning before actually using it.  e.g., use l2_pairs, and before you start your real test, you have every host send a single broadcast ping packet or something (perhaps with a small delay between each one).  This will give each switch a chance to learn each host so that when the real experiment begins, all the tables are ready.  What solution is best depends on what you're trying to do.

**Comment by ppershing** [comment]
I too noticed that POX has some troubles with fattree topos because of loops/discovery timeouts. We worked around this in our experiments by using a custom spanning tree "discovery" which knew exactly the topology from the start and pre-configured the spanning tree accordingly.

**Comment by MurphyMc** [comment]
Yeah.  I actually think the "automatic" discovery isn't the right solution under many circumstances and that giving the controller the topology ahead of time is the right approach.  (I believe this for operational networks as well.)  For that matter, a spanning tree isn't really the right solution for networks with path diversity.  It's all just the "easy" way in some sense.

That said, there have been a number of bugs/problems with POX discovery which have gotten fixed over time (including the push I just made!).  I've also made a number of modifications which have never gotten pushed which can be improvements in some scenarios.  For example, if you're doing experiments that don't involve failures, disabling link timeouts can be useful: discovery is used to find the topology, but then you don't need to worry about links timing out due to overload.  Similarly, you can set the link timeout to be much longer (while keeping the LLDP cycle time the same), and use port events for removing links (with the assumption that link failures are detectable this way; valid in some but not all cases).

(I rarely use any of this stuff anymore, but when I do, it's almost always a combination of loading the topology from a file and then using port events and slow LLDP liveness probes via a modified discovery.  The Mininet-like-thing I use can actually read the same topology files I use in the controller, which makes this pretty convenient.)

**Comment by lordlabakdas** [comment]
Hello,
I get a similar error while on a latest  pulled eel and with a topology with 75 switches and 2 hosts and quite a bit of loops.
I have used the command `./pox.py forwarding.l2_learning openflow.discovery --eat-early-packets openflow.spanning_tree --no-flood --hold-down`
The gist of the log can be found [here](https://gist.github.com/lordlabakdas/229464448629477ab889c266633ff0a9)
I have a topology of 75 switches with quit a bit of loops. As a pre-step to experiments, I am trying to do a `net.pingAll` which would eventually do a ping between the 2 hosts involved. The ping does not go through and I get the same error as is reported earlier here.
`treeping64.py` mininet example pings work without issues.

**Comment by MurphyMc** [comment]
One thing the POX log doesn't show by default which would be helpful is the timestamp.

It looks like discovery is timing out links that it shouldn't, but it's hard to say when this happens and that doesn't help trying to understand why.

Some thoughts:
- With those options, packets definitely won't go through immediately.  There should be a delay while it tries to do the first discovery cycle.  So just keep in mind that it's not like you should be able to fire it up and immediately ping.  It'll take a while to discover.  (That said, this doesn't appear to be your *only* problem anyway, since the log does show links timing out.)
- Pass --link_timeout=20 or =30 (or even something larger) to discovery.  This will mean discovery takes longer to detect all the links, so it'll take longer before you get connectivity, but it makes it somewhat easier for it.  You could also adjust the constant in discovery's send_cycle_time, or just make it return a fixed value.  The default link timeout is 10, so the default send cycle is 5.  You could try keeping the send cycle the same but increasing the timeout.
- Write a simple component that installs a wildcard everything rule with no actions as soon as a switch connects and removes it after the link timeout duration.  Or have it turn off packet-ins by configuring the port with OFPPC_NO_PACKET_IN and then turn them back on.  The idea here is that if the controller is getting hammered with packet-ins and that's keeping it from discovering properly, this may help.
- Make sure there aren't loops *before* the switches connect to POX.  If your switches are set to enable in-band control and/or if they're set to "fail open", and if the switches don't have STP enabled, it's possible there is already a loop and a bunch of replicated packets before they even connect to the controller.  Again, it's hard to tell especially without timestamps, but the large number of repeated messages in the log make it seem like this is a possibility.  I think Mininet's defaults have in-band on and STP off.  Unless you need it, I'd suggest turning in-band off.  And make sure the OVS fail mode is "secure".

**Comment by lordlabakdas** [comment]
My observations:
- I had put in a `time.sleep(20)` just before the pingAll, not sure if it would be helpful. The link timeout that I am using for this case is 50.
- For link timeouts, I set it at a value of 50.
- I have the following code in my `_handle_ConnectionUp` function. Is this behavior consistent with what you had suggested:
```
def _handle_ConnectionUp (self, event):
        log.info("_handle_ConnectionUp Function")
        my_match = of.ofp_match()
        my_match.in_port = None
        my_match.dl_dst = None
        my_match.dl_src = None
        my_match.dl_vlan = None
        my_match.dl_type=None
        my_match.nw_tos=None
        my_match.nw_proto=None
        my_match.nw_src=None
        my_match.nw_dst=None
        my_match.tp_src=None
        my_match.tp_dst=None
        msg=of.ofp_flow_mod(command=of.OFPFC_ADD,
                            hard_timeout=50,
                            match=my_match)
        event.connection.send(msg)
```
- I turned in-band off by `inband=False` in the switch declaration. How do i ensure OVS fail mode is secure?
    - *UPDATE* Ensured that all 75 switches were operating in secure mode.

With the above changes, I still get the same behavior as I used to.

**Comment by MurphyMc** [comment]
Thanks for following up.  Yes, your connection up handler looks fine (though note that an easier way would be not mentioning the match at all -- it defaults to all wildcards).

time.sleep(20) is not going to be sufficient given the entry you installed in your connection up handler.  As a conservative test, try like... time.sleep(70).

And with the new changes, can you try capturing the log again?  In particular, we are interested to see if and when discovery is still timing out links.  It would be helpful to include timestamps.  Here's a starting place for a log format based on an example in the POX manual:

log --format="[%(asctime)s %(name)s %(levelname)s] %(message)s" --datefmt="%H:%M:%S"

Also, is your topology easy to scale?  If so, is there some size at which it works fine and some size at which it starts to fail?  And can you describe your topology (so that one might attempt to recreate your failing experiment)?

**Comment by lordlabakdas** [comment]
The updated log can be found [here](https://gist.github.com/lordlabakdas/177d4bce0bad8fb23d93777b764de0d5). The command used was `./pox.py forwarding.l2_learning openflow.discovery --eat-early-packets openflow.spanning_tree --no-flood --hold-down ext.coronet log --format="[%(asctime)s] %(message)s" --datefmt="%H:%M:%S"`
My topology represents that of an ISP in the US called Coronet and looks like the below:
![image](https://cloud.githubusercontent.com/assets/1252659/24878536/3104c848-1df9-11e7-984e-1c96ef891ae8.png)
The JSON file that possesses the adjacency matrix and the Mininet script that I use can be found [here](https://gist.github.com/lordlabakdas/2aa71ce17d89f091fe0fadd9095e4189)
I will try to work with getting results for scaling the nodes and get back to you.

**Comment by lordlabakdas** [comment]
Some results regarding scaling of nodes here:
10 nodes - 0 pings
20 nodes - pings happened
25 nodes - 0 pings
30 nodes - 0 pings

**Comment by MurphyMc** [comment]
My understanding is that you aren't intentionally having a bunch of dns queries or anything at this point, but looking at the log, there are quite a few.  It quickly balloons from 25 per second right when the "quiet" table entry expires to over 4,000 eight seconds later:
  25 [14:12:21]
 386 [14:12:22]
 480 [14:12:23]
 749 [14:12:24]
 667 [14:12:25]
1045 [14:12:26]
1888 [14:12:27]
3430 [14:12:28]
4311 [14:12:29]

It seems very likely that these are replicated duplicates.  The most logical explanation is that a proper spanning tree is not being established.  The likely reasons would seem to be 1) failure to discover the topology properly, 2) an algorithmic problem computing the spanning tree or, 3) a problem "executing" the tree (that is, something like a problem disabling the appropriate ports).

I'm guessing it's not 1, but this should be easily verifiable by looking at the link detect messages in the beginning of the log and comparing against the adjacency matrix.

2 would be easier to evaluate if we had more info.  Line 108 in spanning_tree.py has an "if" statement that always evaluates to False.  Can you flip it to True and set the POX log level to DEBUG and rerun the experiment?

**Comment by lordlabakdas** [comment]
Reran the case after flipping the if statement to True and the log can be found [here](https://gist.github.com/lordlabakdas/550e21844e6b9ba081583dcb64a6d6d7)

**Comment by lordlabakdas** [comment]
Also, the link detect messages seem to match the number of switches, so it might not be 1)

**Comment by MurphyMc** [comment]
A couple comments...

The beginning of the most recent log file is missing the beginning, which isn't ideal, since I don't think all the link discoveries are in it.  Do you have the beginning?

This is relatively minor, but the log component invocation I gave includes more data than the one in the POX manual: log --format="[%(asctime)s %(name)s %(levelname)s] %(message)s" --datefmt="%H:%M:%S"

I haven't tested it, but I *think* that works.  Not a huge deal, but slightly more convenient for searching.

**Comment by lordlabakdas** [comment]
Yes, I have the begining and updated the [gist](https://gist.github.com/lordlabakdas/550e21844e6b9ba081583dcb64a6d6d7) with the log.
Working on getting the log with the new format.

**Comment by MurphyMc** [comment]
Are you sure the topology is correct?  There seem to be many more links than the figure you posted above would indicate.

**Comment by lordlabakdas** [comment]
The log with the new format using command `./pox.py forwarding.l2_learning openflow.discovery --eat-early-packets openflow.spanning_tree --no-flood --hold-down ext.coronet log --format="[%(asctime)s %(name)s %(levelname)s] %(message)s" --datefmt="%H:%M:%S"` can be found [here](https://gist.github.com/lordlabakdas/4dee01e8b8747062a00c4b0b402cdee1)

**Comment by lordlabakdas** [comment]
I think the topology is correct, a total of 75 switches (and 2 hosts) and 198 links.

**Comment by MurphyMc** [comment]
I think there are actually twice that many (unidirectional) links.  I think maybe you're wiring them up double.  For example:
00-00-00-00-00-71.1 00-00-00-00-00-22.6
00-00-00-00-00-71.3 00-00-00-00-00-22.8

And:
00-00-00-00-00-71.2 00-00-00-00-00-48.4
00-00-00-00-00-71.4 00-00-00-00-00-48.6

I actually have no idea if the existing spanning_tree component deals with this case, but it wouldn't surprise me if it didn't.

(Wiring them up double like that is easy to do if you're iterating the neighbors on each node.  An easy way to fix it is often to skip the link if node2 < node1.)

**Comment by lordlabakdas** [comment]
Sorry my mistake, there should only be 198 unidirectional links. I had considered the whole adj matrix when adding links.
Made the necessary change only considering the upper triangular part of the adj matrix and the new log is provided [here](https://gist.github.com/lordlabakdas/704870c0da727b2a43e37e9e2b8950c7)

**Comment by MurphyMc** [comment]
It looks like we lost the DEBUG level in there, so we don't get to see the tree...

**Comment by lordlabakdas** [comment]
Oops, sorry again. Included the DEBUG log.level and the updated log is [here](https://gist.github.com/lordlabakdas/704870c0da727b2a43e37e9e2b8950c7)

**Comment by MurphyMc** [comment]
So the topology that's discovered is isomorphic with the original coronet graph.  I think possibility 1 is ruled out.

The tree it comes up with does seem to be a tree on the graph... but it's incomplete.  Offhand, this seems like it should cause fewer loops and not more and you seem to be having the problem of more, but either way, this looks like it may be a bug.

Can you try commenting out lines 159-164 in spanning_tree?  You want it to force a tree update on every link event.  The discovery rate is currently set low enough that this won't be a problem.

**Comment by MurphyMc** [comment]
Incidentally, in the grand tradition of such things, I just tried this and it seemed to work just fine.  Whether that's because I'm using a slightly newer POX (now pushed), because I'm not using Mininet, or just because that is always the way of such things, I couldn't possibly guess. :)

**Comment by lordlabakdas** [comment]
I tried commenting out lines 159-164 of `spanning_tree.py` and I am still dont see pings going through.
The log is provided [here](https://gist.github.com/lordlabakdas/ad577bfbee1feeb2896948a7cc1c63c0)

**Comment by ppershing** [comment]
Fyi
We always had problems with discovery on more than let's say 10 switches. Somehow something always times out or a packet is lost or just pox is slow enough to handle this. Our usual solution is to include own discovery component which knows exactly the topology (read it from the same file as mininet)
Peter

-----Original Message-----

**Comment by lordlabakdas** [comment]
@ppershing Would you be willing to provide an example of your custom discovery script for me to look at?

**Comment by MurphyMc** [comment]
@ppershing: I think that's the right approach anyway when possible.

Though the log from @lordlabakdas looked like maybe there's a legit bug.  Unfortunately, recreating it locally wasn't as easy as recreating the topology... 75 switches no problem here.

**Comment by lordlabakdas** [comment]
Some good news :)
I am not sure what changed from the log I posted yesterday to now (other than a reboot), but I am able to ping successfully between the couple of hosts.

**Comment by MurphyMc** [comment]
Well that's a start, at least!

Is your CPU usage super high (indicating a loop)?

**Comment by lordlabakdas** [comment]
Yes, you are right.
A lot of times, during the tests, my laptop freezes and only a hard restart gets it back to working state.

**Comment by MurphyMc** [comment]
Still needs some work and no doubt full of bugs, but give this a try:

https:// gist.github.com/MurphyMc/ 34074ca2336aa1f6d8d9bf716003c703

**Comment by lordlabakdas** [comment]
If I understood right, I used openflow.spanning_forest instead of openflow.spanning_tree. The command I used was `./pox.py forwarding.l2_learning openflow.discovery --eat-early-packets --link_timeout=50 openflow.spanning_forest ext.coronet log --format="[%(asctime)s %(name)s %(levelname)s] %(message)s" --datefmt="%H:%M:%S" log.level --DEBUG`
The log can be found [here](https://gist.github.com/lordlabakdas/328d2ce6bdb7a9d53b0e81a826aa689f)

**Comment by MurphyMc** [comment]
Try just:
./pox.py forwarding.l2_learning openflow.discovery openflow.spanning_forest log --format="[%(asctime)s %(name)s %(levelname)s] %(message)s" --datefmt="%H:%M:%S" log.level --DEBUG

In other words, no custom code.  And make sure you're running the latest POX.

**Comment by lordlabakdas** [comment]
The pings were successful :)

**Comment by MurphyMc** [comment]
And the CPU usage?  Does it get stuck at 100%?

**Comment by lordlabakdas** [comment]
The CPU usage was not high at all....probably averaging around 10%

**Comment by MurphyMc** [comment]
Great.  Let me know if it continues to work for you.  Maybe it's getting to be time to replace the spanning_tree component.

**Comment by lordlabakdas** [comment]
Awesome!
Thanks a lot for all the help.

**Comment by lordlabakdas** [comment]
Sure will keep you updated with any issues I encounter

**Comment by lordlabakdas** [comment]
I am having issues with pings  using spanning_forest on a simple tree topology with a single loop. Given below are the results I see:
With single loop and using spanning_forest: pings do not work
Without single loop and using spanning_forest: pings do not work
without single loop and not using spanning_forest: pings work
I have attached the Mininet test script along with logs for the cases where pings do not work [here](https://gist.github.com/lordlabakdas/eeedd8340241159839b93c74e4ba42a2)

**Comment by MurphyMc** [comment]
From a quick glance at the log, it looks like discovery isn't detecting any links, which would certainly cause problems.

What's your commandline?  Can you remove components and see discovery start/stop working?  e.g., in one of your cases where things don't work, can you try removing any custom forwarding component and try removing the spanning_forest component?  We just want to see discovery mention links being detected.

I guess this could also be due to log config.  I don't remember off the top of my head if order matters, but make sure you're setting the log level first thing, just in case, and launching other components after.

**Comment by lordlabakdas** [comment]
My commandline for the above results was `./pox.py forwarding.l2_learning openflow.discovery openflow.spanning_forest log --format="[%(asctime)s %(name)s %(levelname)s] %(message)s" --datefmt="%H:%M:%S" log.level --DEBUG` 

I then removed `openflow.spanning_forest` and changed the log order(commandline=`./pox.py log --format="[%(asctime)s %(name)s %(levelname)s] %(message)s" --datefmt="%H:%M:%S" log.level --DEBUG openflow.discovery forwarding.l2_learning`) and pings were successful. This is for the case where there are no loops in the network. But when I put back `spanning_forest`(preserving the log order), pings were unsuccessful. There were no other custom components that I was using.

**Comment by MurphyMc** [comment]
Right, without spanning_forest, we'd expect it to work because no links are being disabled.  But did you see link detected events from discovery?

**Comment by lordlabakdas** [comment]
I do not see any links events from discovery. The only events I see are the Installing flow events as given below in both the successful ping and the unsuccessful ping scenarios
```
$ grep "discovery" test.txt
[17:05:34 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-01
[17:05:34 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-03
[17:05:34 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-02
[17:05:34 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-04
[17:05:34 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-05
```

**Comment by MurphyMc** [comment]
The last thing to try for the moment is without l2_learning.  Just `./pox.py log --format="[%(asctime)s %(name)s %(levelname)s] %(message)s" --datefmt="%H:%M:%S" log.level --DEBUG openflow.discovery`.

I'm guessing you aren't going to see any link detected events, which is pretty surprising.  It's not like it hasn't been used on simple tree/linear topologies like this many times.  That needs to be investigated.

**Comment by lordlabakdas** [comment]
I see link detected events now and are pasted below. The entire log is posted [here](https://gist.github.com/lordlabakdas/c35dd3a04274a308dec47885d9ac65e8)
```
[17:23:19 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-01
[17:23:19 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-03
[17:23:19 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-02
[17:23:20 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-04
[17:23:20 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-05
[17:23:20 openflow.discovery INFO] link detected: 00-00-00-00-00-01.1 -> 00-00-00-00-00-02.2
[17:23:21 openflow.discovery INFO] link detected: 00-00-00-00-00-01.2 -> 00-00-00-00-00-03.1
[17:23:21 openflow.discovery INFO] link detected: 00-00-00-00-00-03.1 -> 00-00-00-00-00-01.2
[17:23:22 openflow.discovery INFO] link detected: 00-00-00-00-00-03.2 -> 00-00-00-00-00-05.1
[17:23:22 openflow.discovery INFO] link detected: 00-00-00-00-00-02.1 -> 00-00-00-00-00-04.2
[17:23:23 openflow.discovery INFO] link detected: 00-00-00-00-00-02.2 -> 00-00-00-00-00-01.1
[17:23:24 openflow.discovery INFO] link detected: 00-00-00-00-00-04.2 -> 00-00-00-00-00-02.1
[17:23:24 openflow.discovery INFO] link detected: 00-00-00-00-00-05.1 -> 00-00-00-00-00-03.2
[17:23:19 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-01
[17:23:19 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-03
[17:23:19 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-02
[17:23:20 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-04
[17:23:20 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-05
[17:23:20 openflow.discovery INFO] link detected: 00-00-00-00-00-01.1 -> 00-00-00-00-00-02.2
[17:23:21 openflow.discovery INFO] link detected: 00-00-00-00-00-01.2 -> 00-00-00-00-00-03.1
[17:23:21 openflow.discovery INFO] link detected: 00-00-00-00-00-03.1 -> 00-00-00-00-00-01.2
[17:23:22 openflow.discovery INFO] link detected: 00-00-00-00-00-03.2 -> 00-00-00-00-00-05.1
[17:23:22 openflow.discovery INFO] link detected: 00-00-00-00-00-02.1 -> 00-00-00-00-00-04.2
[17:23:23 openflow.discovery INFO] link detected: 00-00-00-00-00-02.2 -> 00-00-00-00-00-01.1
[17:23:24 openflow.discovery INFO] link detected: 00-00-00-00-00-04.2 -> 00-00-00-00-00-02.1
[17:23:24 openflow.discovery INFO] link detected: 00-00-00-00-00-05.1 -> 00-00-00-00-00-03.2
```

**Comment by MurphyMc** [comment]
Can you confirm that it's a matter of whether l2_learning is included or not?  It could also be like... whether l2_learning is loaded before or after discovery, or when you start the controller up relative to when you start up the switches.  I'm rusty enough at this stuff that I can't be sure what matters and what doesn't. :)

If it seems to be l2_learning that's causing the problem, you should be able to start up with everything except it and spanning_forest will at least build good trees even though forwarding won't work (the current messages all say 0 of 0 links, which is bad... we'd like to see x of y links where x and y are both greater than 0!).

**Comment by lordlabakdas** [comment]
It seems like I see links detected events only when `openflow.discovery` is the only component. Once any other component is added, the link detected events disappear. The order of the components does not seem to matter.
The `spanning_forest` component always shows 0 of 0 links in whatever combination of components it is in, with `l2_learning` and without. 
The only time there were successful pings (that is without a loop in the topology) is when `forwarding.l2_learning openflow.discovery` were used.

The logs for the various cases are given below:

```
#The no loop case which had successful pings
$ ./pox.py log --format="[%(asctime)s %(name)s %(levelname)s] %(message)s" --datefmt="%H:%M:%S" log.level --DEBUG forwarding.l2_learning openflow.discovery > test.txt 2>&1
[17:55:46 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-01
[17:55:46 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-02
[17:55:46 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-03
[17:55:46 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-04
[17:55:46 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-05

# Unsuccessful pings
$ ./pox.py log --format="[%(asctime)s %(name)s %(levelname)s] %(message)s" --datefmt="%H:%M:%S" log.level --DEBUG openflow.discovery > test.txt 2>&1
[17:58:21 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-01
[17:58:21 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-02
[17:58:21 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-04
[17:58:21 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-03
[17:58:21 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-05
[17:58:21 openflow.discovery INFO] link detected: 00-00-00-00-00-01.1 -> 00-00-00-00-00-02.2
[17:58:22 openflow.discovery INFO] link detected: 00-00-00-00-00-01.2 -> 00-00-00-00-00-03.1
[17:58:22 openflow.discovery INFO] link detected: 00-00-00-00-00-02.1 -> 00-00-00-00-00-04.2
[17:58:23 openflow.discovery INFO] link detected: 00-00-00-00-00-02.2 -> 00-00-00-00-00-01.1
[17:58:24 openflow.discovery INFO] link detected: 00-00-00-00-00-04.2 -> 00-00-00-00-00-02.1
[17:58:24 openflow.discovery INFO] link detected: 00-00-00-00-00-03.1 -> 00-00-00-00-00-01.2
[17:58:25 openflow.discovery INFO] link detected: 00-00-00-00-00-03.2 -> 00-00-00-00-00-05.1
[17:58:25 openflow.discovery INFO] link detected: 00-00-00-00-00-05.1 -> 00-00-00-00-00-03.2


#Unsuccessful pings
$ ./pox.py log --format="[%(asctime)s %(name)s %(levelname)s] %(message)s" --datefmt="%H:%M:%S" log.level --DEBUG openflow.discovery openflow.spanning_forest > test.txt 2>&1
[17:59:16 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-02
[17:59:16 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-01
[17:59:16 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-03
[17:59:16 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-04
[17:59:16 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-05
[18:16:43 openflow.spanning_forest DEBUG] Computed spanning forest: 0 of 0 links
[18:16:43 openflow.spanning_forest.00-00-00-00-00-02 INFO] Configured 2 ports
[18:16:43 openflow.spanning_forest DEBUG] Computed spanning forest: 0 of 0 links
[18:16:43 openflow.spanning_forest.00-00-00-00-00-01 INFO] Configured 2 ports
[18:16:43 openflow.spanning_forest DEBUG] Computed spanning forest: 0 of 0 links
[18:16:43 openflow.spanning_forest.00-00-00-00-00-03 INFO] Configured 2 ports
[18:16:43 openflow.spanning_forest DEBUG] Computed spanning forest: 0 of 0 links
[18:16:43 openflow.spanning_forest.00-00-00-00-00-04 INFO] Configured 2 ports
[18:16:43 openflow.spanning_forest DEBUG] Computed spanning forest: 0 of 0 links
[18:16:43 openflow.spanning_forest.00-00-00-00-00-05 INFO] Configured 2 ports
[18:16:49 openflow.spanning_forest DEBUG] Computed spanning forest: 0 of 0 links
[18:16:49 openflow.spanning_forest.00-00-00-00-00-01 INFO] Configured 0 ports
[18:16:49 openflow.spanning_forest.00-00-00-00-00-02 INFO] Configured 0 ports
[18:16:49 openflow.spanning_forest.00-00-00-00-00-03 INFO] Configured 0 ports
[18:16:49 openflow.spanning_forest.00-00-00-00-00-04 INFO] Configured 0 ports
[18:16:49 openflow.spanning_forest DEBUG] Computed spanning forest: 0 of 0 links
[18:16:50 openflow.spanning_forest DEBUG] Computed spanning forest: 0 of 0 links
[18:16:50 openflow.spanning_forest DEBUG] Computed spanning forest: 0 of 0 links
[18:16:50 openflow.spanning_forest DEBUG] Computed spanning forest: 0 of 0 links

# unsuccessful pings
$ ./pox.py log --format="[%(asctime)s %(name)s %(levelname)s] %(message)s" --datefmt="%H:%M:%S" log.level --DEBUG openflow.discovery openflow.spanning_forest forwarding.l2_learning > test.txt 2>&1
[18:10:10 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-02
[18:10:10 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-01
[18:10:10 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-03
[18:10:10 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-04
[18:10:10 openflow.discovery DEBUG] Installing flow for 00-00-00-00-00-05
[18:15:27 openflow.spanning_forest DEBUG] Computed spanning forest: 0 of 0 links
[18:15:27 openflow.spanning_forest.00-00-00-00-00-02 INFO] Configured 2 ports
[18:15:27 openflow.spanning_forest DEBUG] Computed spanning forest: 0 of 0 links
[18:15:27 openflow.spanning_forest.00-00-00-00-00-01 INFO] Configured 2 ports
[18:15:27 openflow.spanning_forest DEBUG] Computed spanning forest: 0 of 0 links
[18:15:27 openflow.spanning_forest.00-00-00-00-00-03 INFO] Configured 2 ports
[18:15:27 openflow.spanning_forest DEBUG] Computed spanning forest: 0 of 0 links
[18:15:27 openflow.spanning_forest.00-00-00-00-00-05 INFO] Configured 2 ports
[18:15:27 openflow.spanning_forest DEBUG] Computed spanning forest: 0 of 0 links
[18:15:27 openflow.spanning_forest.00-00-00-00-00-04 INFO] Configured 2 ports
[18:15:33 openflow.spanning_forest DEBUG] Computed spanning forest: 0 of 0 links
[18:15:33 openflow.spanning_forest.00-00-00-00-00-01 INFO] Configured 0 ports
[18:15:33 openflow.spanning_forest.00-00-00-00-00-02 INFO] Configured 0 ports
[18:15:33 openflow.spanning_forest.00-00-00-00-00-03 INFO] Configured 0 ports
[18:15:33 openflow.spanning_forest.00-00-00-00-00-04 INFO] Configured 0 ports
[18:15:33 openflow.spanning_forest DEBUG] Computed spanning forest: 0 of 0 links
[18:15:33 openflow.spanning_forest DEBUG] Computed spanning forest: 0 of 0 links
[18:15:33 openflow.spanning_forest DEBUG] Computed spanning forest: 0 of 0 links
[18:15:33 openflow.spanning_forest DEBUG] Computed spanning forest: 0 of 0 links
```

**Comment by MurphyMc** [comment]
That seems pretty strange.  Unfortunately, I don't have time to investigate until after the 10th...

**Comment by humikahout** [comment]
I am running pox with mininet Some of the errors I get ERROR:openflow.of_01:[00-00-00-00-00-20 457] OpenFlow Error: [00-00-00-00-00-20 457] Error: header: [00-00-00-00-00-20 457] Error: version: 1 [00-00-00-00-00-20 457] Error: type: 1 (OFPT_ERROR) [00-00-00-00-00-20 457] Error: length: 36 [00-00-00-00-00-20 457] Error: xid: 112392 [00-00-00-00-00-20 457] Error: type: OFPET_BAD_REQUEST (1) [00-00-00-00-00-20 457] Error: code: OFPBRC_BUFFER_UNKNOWN (8) [00-00-00-00-00-20 457] Error: datalen: 24 [00-00-00-00-00-20 457] Error: 0000: 01 0d 00 18 00 01 b7 08 00 00 03 4f 0000 08 |...........O....| [00-00-00-00-00-20 457] Error: 0010: 00 00 00 08 ff fb 00 00 |........ | And network topology code that i run with mininet from mininet.cli import CLI from mininet.net import Mininet from mininet.node import Controller, RemoteController, OVSController from mininet.link import TCLink from mininet.util import dumpNodeConnections from mininet.log import lg if _name_ == '__main__': lg.setLogLevel('info') net = Mininet( controller=RemoteController, link=TCLink) #listenPort is necessary for adding flow entry net.addController( 'c0', controller=RemoteController, ip='192.168.10.7', port=6633 ) #HOST constructor H1 = net.addHost('h1',mac='aa:aa:aa:aa:aa:01') #VIDEO STREAMING SERVER H2 = net.addHost('h2',mac='aa:aa:aa:aa:aa:02') #MEDICAL SERVER H3 = net.addHost('h3',mac='aa:aa:aa:aa:aa:03') #MULTIMEDIA CLIENT H4 = net.addHost('h4',mac='aa:aa:aa:aa:aa:04') #MEDICAL CLIENT #SWITCH constructor; S1 = net.addSwitch('s1', listenPort=6634) S2 = net.addSwitch('s2', listenPort=6635) S3 = net.addSwitch('s3', listenPort=6636) S4 = net.addSwitch('s4', listenPort=6637) S5 = net.addSwitch('s5', listenPort=6638) S6 = net.addSwitch('s6', listenPort=6639) #LINK constructor net.addLink(H3, S1, delay='1ms', loss=0, bw=10) net.addLink(H4, S4, delay='1ms', loss=0, bw=10) net.addLink(S1, S2, delay='1ms', loss=0, bw=10) net.addLink(S1, S5, delay='1ms', loss=0, bw=10) net.addLink(S1, S4, delay='1ms', loss=0, bw=10) net.addLink(S4, S2, delay='1ms', loss=0, bw=10) net.addLink(S4, S5, delay='1ms', loss=0, bw=10) net.addLink(S2, S3, delay='1ms', loss=0, bw=10) net.addLink(S2, S6, delay='1ms', loss=0, bw=10) net.addLink(S2, S5, delay='1ms', loss=0, bw=10) net.addLink(S5, S3, delay='1ms', loss=0, bw=10) net.addLink(S5, S6, delay='1ms', loss=0, bw=10) net.addLink(S3, H1, delay='1ms', loss=0, bw=10) net.addLink(S3, S6, delay='1ms', loss=0, bw=10) net.addLink(S6, H2, delay='1ms', loss=0, bw=10) # Start network net.start() CLI( net ) print "*** Dumping host connections" dumpNodeConnections(net.hosts) print "*** Testing network connectivity" net.pingALL() print " host are running " print " type 'exit' or control-D to shut down network " net.stop()﻿

**Comment by MurphyMc** [comment]
@humikahout, the above message is awfully hard to read.

Are you sure this is related to this issue?  Does your topology have a loop?

What POX components are you running?

**Comment by EmanSalameh** [comment]
I have the same error, but I didn't run fattree topology. I run my topology with two pox controllers
ERROR:openflow.of_01:[00-00-00-00-00-01 4] OpenFlow Error:
[00-00-00-00-00-01 4] Error: header: 
[00-00-00-00-00-01 4] Error:   version: 1
[00-00-00-00-00-01 4] Error:   type:    1 (OFPT_ERROR)
[00-00-00-00-00-01 4] Error:   length:  76
[00-00-00-00-00-01 4] Error:   xid:     1034
[00-00-00-00-00-01 4] Error: type: OFPET_BAD_REQUEST (1)
[00-00-00-00-00-01 4] Error: code: OFPBRC_BUFFER_UNKNOWN (8)
[00-00-00-00-00-01 4] Error: datalen: 64
[00-00-00-00-00-01 4] Error: 0000: 01 0e 00 50 00 00 04 0a 

I want to know what is the reason of this error?
my topology is:
`def myNetwork():

    net = Mininet( topo=None, build=False, link=TCLink, controller=RemoteController)

    linkopts = dict(bw=100, delay='1us', loss=0)
    info( '*** Adding controller ***\n' )
    controller1=net.addController(name='controller1', controller=RemoteController, protocol='tcp', port=6633)
    controller2=net.addController( name='controller2', controller=RemoteController, protocol='tcp', port=6644)
    

    info( '*** Add switches ***\n')
    
    switch1 = net.addSwitch('switch1', cls=OVSKernelSwitch)
    switch2 = net.addSwitch('switch2', cls=OVSKernelSwitch)
    net.addLink(switch1, switch2, **linkopts)

    #switch1.cmd ('ovs-vsctl set Bridge switch1 protocols=OpenFlow10')
    #switch2.cmd ('ovs-vsctl set Bridge switch2 protocols=OpenFlow10')

    server1 = net.addHost('server1', cls=Node, ip='10.0.0.1', mac='00:00:00:00:00:01')
    net.addLink(switch1, server1, **linkopts)

    server2 = net.addHost('server2', cls=Node, ip='10.0.0.2', mac= '00:00:00:00:00:02')
    net.addLink(switch2, server2, **linkopts)
    service = net.addHost('service', cls=Node, ip='10.0.0.3', mac= '00:00:00:00:00:03')
    net.addLink(switch1, service, **linkopts)

    info( '*** Add hosts\n')
    for h in range(0,31):#
             client = net.addHost('client%s' % (h + 1),cls=Node, ip='10.0.0.%s'%(h+4), mac= '00:00:00:00:00:0%s'%(h+4))
             net.addLink(client, switch1, **linkopts)

    for h in range(31,64):#(2,4)
             client = net.addHost('client%s' % (h + 1), cls=Node, ip='10.0.0.%s'%(h+4), mac= '00:00:00:00:00:0%s'%(h+4))
             net.addLink(client, switch2, **linkopts)



    info( '*** Starting network ***\n')
    net.build()
    net. staticArp()
    controller1.start()
    switch1.start([controller1])
    controller2.start()
    switch2.start([controller2])

 
    info( '\n*** Starting web servers ***\n')
    server1 = net.get('server1')
    server1.cmd('iperf -s -p 80 &')
    server2 = net.get('server2')
    server2.cmd('iperf -s -p 80 &')

    time.sleep(20)    
    #Start the simulation
    info( '*** Starting the simulation ***\n')
    start_time = time.time()
    finish_time = time.time() - start_time

    cli = 1
    cli2 = 32
    while (finish_time <=20): # run for 300 seconds (simulation time)
       
       init_time = time.time()
       sec_time = time.time() -init_time
       while (sec_time <= 1):
          for flow in range(0,3):   #Generate 6 flows/s on switch1
             #rnd = random.randint(1, 31)
             host =  net.get('client%s' % (cli))
             host.cmd('iperf  -c 10.0.0.3 -p 80 -n 51K -N  &')
             cli+=1
             if cli >= 32: 
                cli = 1

          for flow in range(0,2):   #Generate 4 flows/s on switch2
             #rnd = random.randint(32, 64)
             host =  net.get('client%s' % (cli2))
             host.cmd('iperf  -c 10.0.0.3  -p 80  -n 51K -N  &')
             cli2+=1
             if cli2 >= 64:
                cli2 = 32

          # End of 1 second
          sec_time = time.time() - init_time
          print "sec_time = " , sec_time
       time.sleep(2)
       #End of simulation
       finish_time = time.time() - start_time
    info( '*** End of simulation ***\n')

    CLI(net)
    net.stop()

if __name__ == '__main__':
    setLogLevel( 'info' )
    myNetwork()`

**Comment by MurphyMc** [comment]
I doubt this is the same issue.  You should open a new one.  And please describe what you're doing and why and what POX components you're running and what your topology looks like.  It looks like it's just two switches and some hosts, but I am not going read your Mininet code carefully.

Also, have you tried eliminating variables first?  What happens if you only run a single controller?

**Comment by humikahout** [comment]
Can u send me ur email address i will email u my custom built topology and bellmanford algorithm and also the screen shots of output than u easilly understand

**Comment by MurphyMc** [comment]
This issue has been idle for a while and has several different things going on.  And I'm hopeful that the new spanning_forest in fangtooth is going to resolve at least some of them.  So I'm going to close it.

If you find this issue and think you have something that relates, I'd say open a new issue and reference this one in it.

**Comment by ghost** [comment]
Some of the errors I also get are:
DEBUG:iplb.d4-ca-6d-b6-c7-ac|1:Directing traffic to 192.168.10.4
ERROR:openflow.of_01:[d4-ca-6d-b6-c7-ac|1 1] OpenFlow Error:
[00-00-00-00-00-20 457] Error: header:
[00-00-00-00-00-20 457] Error: version: 1
[00-00-00-00-00-20 457] Error: type: 1 (OFPT_ERROR)
[00-00-00-00-00-20 457] Error: length: 116
[00-00-00-00-00-20 457] Error: xid: 13
[00-00-00-00-00-20 457] Error: type: OFPET_BAD_REQUEST (3)
[00-00-00-00-00-20 457] Error: code: OFPBRC_BUFFER_UNKNOWN (0)
[00-00-00-00-00-20 457] Error: datalen: 104
[00-00-00-00-00-20 457] Error: 0000: 01 0d 00 18 00 01 b7 08 00 00 03 4f 0000 08 |...........O....|
[00-00-00-00-00-20 457] Error: 0010: 00 00 00 08 ff fb 00 00 |........ |

why???

**Comment by MurphyMc** [comment]
It's impossible to say with the information given, and this issue has unrelated things going on.

Please open a new issue and fully describe what you're doing (e.g., which version of POX, which components, what your topology is like, what type of switches, etc.) and what problem you're having.

**Comment by ghost** [comment]
Some of the errors I also get are:
DEBUG:iplb.d4-ca-6d-b6-c7-ac|1:Directing traffic to 192.168.10.4
ERROR:openflow.of_01:[d4-ca-6d-b6-c7-ac|1 1] OpenFlow Error:
[00-00-00-00-00-20 457] Error: header:
[00-00-00-00-00-20 457] Error: version: 1
[00-00-00-00-00-20 457] Error: type: 1 (OFPT_ERROR)
[00-00-00-00-00-20 457] Error: length: 116
[00-00-00-00-00-20 457] Error: xid: 13
[00-00-00-00-00-20 457] Error: type: OFPET_BAD_REQUEST (3)
[00-00-00-00-00-20 457] Error: code: OFPBRC_BUFFER_UNKNOWN (0)
[00-00-00-00-00-20 457] Error: datalen: 104
[00-00-00-00-00-20 457] Error: 0000: 01 0d 00 18 00 01 b7 08 00 00 03 4f 0000 08 |...........O....|
[00-00-00-00-00-20 457] Error: 0010: 00 00 00 08 ff fb 00 00 |........ |

i used POX version eel and using Mikrotik RB951-2n.
